### PR TITLE
fix(test): JUnit reporter includes file, line and column attributes

### LINKED ISSUE #20156: JUnit: Include full test location
The JUnit `testcase` attribute supports providing the file path and line number of a test which can help with debugging by taking developers to the exact test. 
```xml
<testcase name="testCase1" classname="Tests.Registration" assertions="2" time="2.436" file="tests/registration.code" line="24" />
```

### PR DESCRIPTION
**Pull request by armichaud**
Closes #20156

### CODE REVIEW

**Review by bartlomieju — COMMENTED** [review]
That looks good, could you please add an integration test that asserts the output is as expected? You can add one in `cli/tests/integration/test_tests.rs` copying `itest!(pass...` and using `--junit` flag.

**Comment by armichaud** [comment]
@bartlomieju I'm going to commit a unit test, but it's not passing consistently because the `<testcase>` tag has a `time` property whose value reflects the amount of time any given test took to run, and is therefore not consistent. I've got to pause for the night, but tomorrow I'll look into if there's a way that iTest can match patterns within outputs. If you have any suggestions in the meantime, let me know.

File: cli/tests/integration/test_tests.rs [hunk 1]
```diff
@@ -267,6 +267,11 @@ itest!(exit_sanitizer {
   exit_code: 1,
 });
 
+itest!(junit {
+  args: "test --reporter junit test/junit.ts",
+  output: "test/junit.out",
```
Comment (bartlomieju) [inline cli/tests/integration/test_tests.rs:272]:
```suggestion
  output: "test/pass.junit.out",
```

File: cli/tests/testdata/test/junit.out
Comment (bartlomieju) [inline cli/tests/testdata/test/junit.out:1]: Let's rename to `pass.junit.out`

File: cli/tests/testdata/test/junit.out [hunk 2]
```diff
@@ -0,0 +1,118 @@
+Check file:///Users/alexanderrodionmichaud/code/deno/cli/tests/testdata/test/junit.ts
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuites name="deno test" tests="16" failures="0" errors="0" time="0.136">
+    <testsuite name="file:///Users/alexanderrodionmichaud/code/deno/cli/tests/testdata/test/junit.ts" tests="16" disabled="0" errors="0" failures="0">
```
Comment (bartlomieju) [inline cli/tests/testdata/test/junit.out:4]: @skycoop should we use file URLs here or strip the prefix?

**Review by bartlomieju — COMMENTED** [review]
Looks good @armichaud, you can use `[WILDCARD]` to match any pattern and remove flakiness from the tests.

@skycoop can you take a look if this is what you expect?

**Comment by bartlomieju** [comment]
@armichaud I think I made a mistake, we should probably use `extra` field (https://docs.rs/quick-junit/latest/quick_junit/struct.TestCase.html#structfield.extra) instead of properties.

File: cli/tests/testdata/test/junit.out (hunk 2 above)
Comment (armichaud) [reply cli/tests/testdata/test/junit.out:4]: Oh good catch we should definitely not assert that it matches my filesystem 😅

**Comment by armichaud** [comment]
> @armichaud I think I made a mistake, we should probably use `extra` field (https://docs.rs/quick-junit/latest/quick_junit/struct.TestCase.html#structfield.extra) instead of properties.

Interesting, yeah based on the `environment variables` example they give for properties, I think you're right that extras would be more appropriate. I'll update accordingly.

**Comment by armichaud** [comment]
@bartlomieju should be good to go now

File: cli/tests/testdata/test/pass.junit.out [hunk 3]
```diff
@@ -0,0 +1,38 @@
+Check [WILDCARD]/testdata/test/pass.ts
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuites name="deno test" tests="16" failures="0" errors="0" time="[WILDCARD]">
+    <testsuite name="file:///Users/alexanderrodionmichaud/code/deno/cli/tests/testdata/test/pass.ts" tests="16" disabled="0" errors="0" failures="0">
+        <testcase name="test 0" time="[WILDCARD]" filename="file:///Users/alexanderrodionmichaud/code/deno/cli/tests/testdata/test/pass.ts" line="1" col="6">
```
Comment (bartlomieju) [inline cli/tests/testdata/test/pass.junit.out:5]: These should still be updated to use `[WILDCARD]`

File: cli/tests/testdata/test/pass.junit.out [hunk 4]
```diff
@@ -0,0 +1,38 @@
+Check [WILDCARD]/testdata/test/pass.ts
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuites name="deno test" tests="16" failures="0" errors="0" time="[WILDCARD]">
+    <testsuite name="file:///Users/alexanderrodionmichaud/code/deno/cli/tests/testdata/test/pass.ts" tests="16" disabled="0" errors="0" failures="0">
```
Comment (bartlomieju) [inline cli/tests/testdata/test/pass.junit.out:4]: Similarly here

File: cli/tools/test/reporters/junit.rs [hunk 5]
```diff
@@ -40,13 +40,23 @@ impl JunitTestReporter {
...
-        description.name.clone(),
-        quick_junit::TestCaseStatus::skipped(),
-      ),
+    let mut case = quick_junit::TestCase::new(
+      description.name.clone(),
+      quick_junit::TestCaseStatus::skipped(),
     );
+    case.extra.insert(
+      String::from("filename"),
+      description.location.file_name.clone(),
```
Comment (bartlomieju) [inline cli/tools/test/reporters/junit.rs:49]: I also think we should strip out `file://` prefix from these. I haven't found an example of JUnit report that used file URL. That should be something like `.strip_prefix("file://")`

**Comment by bartlomieju** [comment]
Thanks @armichaud, awesome work! I'll change the title of this PR to be a fix and merge it before the next patch release.

**Comment by armichaud** [comment]
@bartlomieju Thanks for your help on my first open source PR! I look forward to contributing more to Deno.

**Review by bartlomieju — APPROVED** [review]
LGTM, nice work!

**Pull request merged**